### DISCUSSION TITLE
How to make await calls in `init()`?

### DISCUSSION THREAD

**Discussion by eric-hansen** [alpinejs/alpine#4410] [1. Help] [answered]
I've tried Googling this for a while and attempting a few different things, but nothing seems to work.

I have some `Promises` I need to resolve for `x-init`, so I created a `x-data="{ init() { .... } }"` on the `div`.

The things I can remember trying for this are:
* `async init () { ... }`
* `init: async function () { ... }`
* creating an async function and then setting `x-init="async someFunc"`

Nothing seems to work, so I'm not sure if I'm doing this wrong or if it's just not possible.  If it's not, then how can I resolve the promises without suffering from callback hell?

**Comment by SimoTod** [comment]
Can you share the code to show what you are trying to do. Init functions support await but I suspect you mean something else (like you want other directives to wait for init)?

Simple example of await in x-init: https://codepen.io/SimoTod/pen/jOgLjqp

**Reply by eric-hansen** [reply]
Here's what I'm trying to do right now (Alpine + Livewire/Laravel):

```
<div
    x-data="{
        map,
        markers: [{
            lat: @js($location->latitude()),
            lng: @js($location->longitude()),
        }],
        markerHandler,
        placeMarkers () {
            console.log(this.markerHandler);
        },
        initMap() {
            let m;
            let bounds;
            let mapEl = this.$refs.map;

            let center = { ...this.markers[0] };

            google.maps.importLibrary('maps').then((MapsLibrary) => {
                let Map = MapsLibrary.Map;

                this.map = new Map(mapEl, {
                    center: center,
                    zoom: 8,
                    mapId: 'demoIdMapYall',
                    scaleControl: true
                });

                let that = this;

                google.maps.importLibrary('marker').then((MarkerLibrary) => {
                    that.markerHandler = MarkerLibrary.AdvancedMarkerElement;
console.log('that.markerHandler', that.markerHandler);
                    placeMarkers();
                });
            });
       }
    }"
    x-init="await initMap()"
wire:ignore>
</div>
```

And I get this error:

```
Uncaught ReferenceError: initMap is not defined
```

This is at the `await initMap()` call.  I know I don't need the callback hell in `initMap()` but that's how it was working and so I just did the above to see if I could get past the "not defined" error I keep having.

I also get the same error when I do `async initMap()` in `x-data` (which I forgot to add originally).

**Comment by ekwoka** [comment]
just making init async works fine.

**Reply by eric-hansen** [reply]
How so?  I'm wondering if I'm doing it wrong somehow (I'm more of a backend dev but need to do this for Google Maps integration).

**Reply by eric-hansen** [reply]
For example, here's what I'm trying to do right now (Alpine + Livewire/Laravel):

```
<div
    x-data="{
        map,
        markers: [{
            lat: @js($location->latitude()),
            lng: @js($location->longitude()),
        }],
        markerHandler,
        placeMarkers () {
            console.log(this.markerHandler);
        },
        initMap() {
            let m;
            let bounds;
            let mapEl = this.$refs.map;

            let center = { ...this.markers[0] };

            google.maps.importLibrary('maps').then((MapsLibrary) => {
                let Map = MapsLibrary.Map;

                this.map = new Map(mapEl, {
                    center: center,
                    zoom: 8,
                    mapId: 'demoIdMapYall',
                    scaleControl: true
                });

                let that = this;

                google.maps.importLibrary('marker').then((MarkerLibrary) => {
                    that.markerHandler = MarkerLibrary.AdvancedMarkerElement;
console.log('that.markerHandler', that.markerHandler);
                    placeMarkers();
                });
            });
       }
    }"
    x-init="await initMap()"
wire:ignore>
</div>
```

And I get this error:

```
Uncaught ReferenceError: initMap is not defined
```

This is at the `await initMap()` call.  I know I don't need the callback hell in `initMap()` but that's how it was working and so I just did the above to see if I could get past the "not defined" error I keep having.

I also get the same error when I do `async initMap()` in `x-data` (which I forgot to add originally).

**Reply by ekwoka** [reply]
how about just `async init() { //do stuff }`?

you don't need the x-init either.

strange that its saying it's not defined though...

is the x-data failing to evaluate?

when I test it the x-data is failing. That's the problem. Once that fails, the init fails (of course)

**Reply by eric-hansen** [reply, marked as answer]
That would give me the same error.

`x-data` isn't failing to evaluate for me, but I'm also using Livewire/Laravel `@js()` to transform data, which may be why it fails for you.

I did get it working though, with this being the final result:

```
<div
    x-data="{
        map,
        markers: [{
            lat: @js($location->latitude()),
            lng: @js($location->longitude()),
        }],
        markerHandler,
        placeMarkers () {
            // console.log(this.markerHandler);
        },
        init: async function () {
            let m;
            let bounds;
            let mapEl = this.$refs.map;

            let center = { ...this.markers[0] };

            const { Map } = await google.maps.importLibrary('maps');

            this.map = new Map(mapEl, {
                center: center,
                zoom: 8,
                mapId: 'demoIdMapYall',
                scaleControl: true
            });

            const { AdvancedMarkerElement } = await google.maps.importLibrary('marker');

            // this.markerHandler = AdvancedMarkerElement;

            placeMarkers();
       }
    }"
wire:ignore>
</div>
```

Basically I add define it as `init: async function () { // ... }`

**Reply by ekwoka** [reply]
I had removed that part, the map and stuff were failing.

once I fixed those references, the init worked fine.

> init: async function () { // ... }

You can just do `async init() { }`

**Comment by Stef16Robbe** [comment]
I'm struggling with a similar issue. I have a simple use case; retrieve JSON data from an API, and render rows in a table based on it. Here's a snippet:

```plain
<table
    class="table table-lg table-pin-rows table-pin-cols"
    x-data="{
      restaurants: [],
    }"
    x-init="let response = await fetch('/api/restaurants'); this.restaurants = await response.json(); console.log(this.restaurants)"
  >
    <thead>
      <tr>
        <td width="10%">Name</td>
        <td width="10%">Distance</td>
        <td width="10%">Outdoor</td>
        <td width="10%">Rating</td>
        <th width="10%"></th>
      </tr>
    </thead>
    <tbody>
      <template x-init="console.log(restaurants)" x-for="(restaurant, index) in restaurants" :key="index">
        <tr>
          <td><span x-text="restaurant.name"></span></td>
          <td><span x-text="restaurant.distance_km"></span>km</td>
          <td><span x-text="restaurant.outdoor"></span></td>
...
```
So the `console.log()` in the `x-init` of the `table` shows the array as expected. But then no `tr`'s are ever rendered. I added a `console.log()` in the `x-init` below it but that returns a `Proxy` that's empty:

```
Proxy { <target>: [], <handler>: {…} }
​
<target>: Array []
​​
length: 0
​​
<prototype>: Array []
​
<handler>: Object { get: cn(n, i, o), set: Oi(r, n, i, o), deleteProperty: Ci(e, t), … }
​​
deleteProperty: function Ci(e, t)
​​
get: function cn(n, i, o)
​​
has: function Ti(e, t)
​​
ownKeys: function Ri(e)
​​
set: function Oi(r, n, i, o)
​​
<prototype>: Object { … }
cdn.min.js line 5 > AsyncFunction:3:40
```

Not sure what's going wrong here.

**Reply by SimoTod** [reply]
Shouldn't it be 
```
x-init="let response = await fetch('/api/restaurants'); restaurants = await response.json();"
```

**Reply by Stef16Robbe** [reply]
I thought I had tried that already before... Weekend is close luckily.. 🥲

Thanks so much for the quick reply!
